sdc-plugin/clocks.cc

```diff
@@ -19,220 +19,104 @@
-	    AddClock(clock.Name(), aliases, clock.Period(),
-	                  clock.RisingEdge(), clock.FallingEdge());
-	}
-    }
-#ifdef SDC_DEBUG
-    log("Finish natural clock propagation\n\n");
-#endif
+float Clock::RisingEdge(RTLIL::Wire* clock_wire) {
+    return Waveform(clock_wire).first;
 }
 
-void Clocks::Propagate(BufferPropagation* pass) {
-#ifdef SDC_DEBUG
-    log("Start buffer clock propagation\n");
-    log("IBUF pass\n");
-#endif
-    std::vector<Clock> clocks(clocks_);
-    for (auto clock : clocks) {
-#ifdef SDC_DEBUG
-	log("Processing clock %s\n", clock.Name().c_str());
-#endif
-	PropagateThroughBuffer(pass, clock, IBuf());
-    }
-#ifdef SDC_DEBUG
-    log("BUFG pass\n");
-#endif
-    clocks = clocks_;
-    for (auto clock : clocks) {
-#ifdef SDC_DEBUG
-	log("Processing clock %s\n", clock.Name().c_str());
-#endif
-	PropagateThroughBuffer(pass, clock, Bufg());
-    }
-#ifdef SDC_DEBUG
-    log("Finish buffer clock propagation\n\n");
-#endif
+float Clock::FallingEdge(RTLIL::Wire* clock_wire) {
+    return Waveform(clock_wire).second;
 }
 
-void Clocks::Propagate(ClockDividerPropagation* pass) {
-#ifdef SDC_DEBUG
-    log("Start clock divider clock propagation\n");
-#endif
-    for (auto clock : clocks_) {
-#ifdef SDC_DEBUG
-	log("Processing clock %s\n", clock.Name().c_str());
-#endif
-	auto pll_clocks =
-	    pass->FindSinkClocksForCellType(clock, "PLLE2_ADV");
-	for (auto pll_clock : pll_clocks) {
-#ifdef SDC_DEBUG
-	    log("PLL clock: %s\n", pll_clock.Name().c_str());
-#endif
-	    AddClock(pll_clock);
-	    PropagateThroughBuffer(pass, pll_clock, Bufg());
-	}
+std::string Clock::Name(RTLIL::Wire* clock_wire) {
+    if (clock_wire->has_attribute(RTLIL::escape_id("NAME"))) {
+	return clock_wire->get_string_attribute(RTLIL::escape_id("NAME"));
     }
-#ifdef SDC_DEBUG
-    log("Finish clock divider clock propagation\n\n");
-#endif
+    return WireName(clock_wire);
 }
 
-void Clocks::PropagateThroughBuffer(Propagation* pass, Clock& clock,
-                                    Buffer buffer) {
-    auto clock_wires = clock.GetClockWires();
-    for (auto clock_wire : clock_wires) {
-#ifdef SDC_DEBUG
-	log("Clock wire %s\n", RTLIL::unescape_id(clock_wire->name).c_str());
-#endif
-	auto buf_wires = pass->FindSinkWiresForCellType(clock_wire, buffer.name,
-	                                                buffer.output);
-	int path_delay(0);
-	for (auto wire : buf_wires) {
-#ifdef SDC_DEBUG
-	    log("%s wire: %s\n", buffer.name.c_str(),
-	        RTLIL::unescape_id(wire->name).c_str());
-#endif
-	    path_delay += buffer.delay;
-	    AddClock(RTLIL::unescape_id(wire->name), wire, clock.Period(),
-	             clock.RisingEdge() + path_delay,
-	             clock.FallingEdge() + path_delay);
-	}
+std::string Clock::WireName(RTLIL::Wire* wire) {
+    if (!wire) {
+	return std::string();
     }
+    std::string wire_name(RTLIL::unescape_id(wire->name));
+    return std::regex_replace(wire_name, std::regex{"\\$"}, "\\$");
```

Comment:
Can you explain what this regex is doing?  Because at first blush, this doesn't do anything?

Reply:
I will add an comment, but in brief it makes sure we escape `$` in the output sdc e.g. we get
`\$auto\$clkbufmap.cc:247:execute\$1829`
instead of
`$auto$clkbufmap.cc:247:execute$1829`

Reply:
Would it be possible to factor this into a small function, and you write some unit tests for that small function?  Thanks

Reply:
Added googletest and some Makefile template to perform unit tests.

Reply:
Thanks